presto-orc/src/main/java/com/facebook/presto/orc/reader/TimestampSelectiveStreamReader.java

```diff
@@ -0,0 +1,430 @@
+import org.joda.time.DateTime;
+import org.joda.time.DateTimeZone;
+import org.openjdk.jol.info.ClassLayout;
+
+import javax.annotation.Nullable;
+
+import java.io.IOException;
+import java.util.List;
+import java.util.Optional;
+
+import static com.facebook.presto.orc.metadata.Stream.StreamKind.DATA;
+import static com.facebook.presto.orc.metadata.Stream.StreamKind.PRESENT;
+import static com.facebook.presto.orc.metadata.Stream.StreamKind.SECONDARY;
+import static com.facebook.presto.orc.reader.TimestampUtils.decodeTimestamp;
+import static com.facebook.presto.orc.stream.MissingInputStreamSource.missingStreamSource;
+import static com.facebook.presto.spi.type.TimestampType.TIMESTAMP;
+import static com.google.common.base.Preconditions.checkArgument;
+import static com.google.common.base.Preconditions.checkState;
+import static io.airlift.slice.SizeOf.sizeOf;
+import static java.util.Objects.requireNonNull;
+
+public class TimestampSelectiveStreamReader
+        implements SelectiveStreamReader
+{
+    private static final int INSTANCE_SIZE = ClassLayout.parseClass(TimestampSelectiveStreamReader.class).instanceSize();
+    private static final Block NULL_BLOCK = TIMESTAMP.createBlockBuilder(null, 1).appendNull().build();
+
+    private final StreamDescriptor streamDescriptor;
+    private final TupleDomainFilter filter;
+    private final boolean nullsAllowed;
+    private final boolean outputRequired;
+    private final LocalMemoryContext systemMemoryContext;
+    private final long baseTimestampInSeconds;
+
+    private InputStreamSource<BooleanInputStream> presentStreamSource = missingStreamSource(BooleanInputStream.class);
+    private InputStreamSource<LongInputStream> secondsStreamSource = missingStreamSource(LongInputStream.class);
+    private InputStreamSource<LongInputStream> nanosStreamSource = missingStreamSource(LongInputStream.class);
+
+    @Nullable
+    private BooleanInputStream presentStream;
+    private LongInputStream secondsStream;
+    private LongInputStream nanosStream;
+    private boolean rowGroupOpen;
+
+    private int readOffset;
+    @Nullable
+    private long[] values;
+    @Nullable
+    private boolean[] nulls;
+    @Nullable
+    private int[] outputPositions;
+    private int outputPositionCount;
+    private boolean allNulls;
+    private boolean valuesInUse;
+
+    public TimestampSelectiveStreamReader(
+            StreamDescriptor streamDescriptor,
+            Optional<TupleDomainFilter> filter,
+            DateTimeZone hiveStorageTimeZone,
+            boolean outputRequired,
+            LocalMemoryContext systemMemoryContext)
+    {
+        this.streamDescriptor = requireNonNull(streamDescriptor, "streamDescriptor is null");
+        this.filter = filter.orElse(null);
+        this.outputRequired = outputRequired;
+        this.systemMemoryContext = requireNonNull(systemMemoryContext, "systemMemoryContext is null");
+        this.nullsAllowed = this.filter == null || this.filter.testNull();
```

Comment:
This logic needs to be updated to support non-deterministic filters. See BooleanSelectiveStreamReader.